add heartbeat reporting

ticoann opened the pull request:
tested in my vm.  need to update with https://github.com/dmwm/deployment/pull/437

ticoann committed:
add heartbeat reporting

handle wmstats thread

change wmstats url config

fix wmstats url

fix as alan's suggestion

change the package location

fix gq monitoring config

fix typo

fix typo

move import to the top

ticoann committed:
add unittest

ticoann commented:
test this please

ticoann commented:
test this please

ticoann commented:
@amaltaro, Alan please review. It seems to be working in my vm test.

amaltaro commented on src/python/WMCore/Services/LogDB/LogDB.py:
```diff
@@ -126,5 +127,55 @@ def cleanup(self, thr, backend='local'):
+            # wmstats thread can run in multiple boxed. 
+            if self.identifier == identifier or identifier.startswith(self.identifier):
+                #this will handle wmstats DataCacheUpdate thread with multiple machine
+                postfix = identifier.replace(self.identifier, "")
```
I think I got lost in this review. First, why checking if identifier startswith ... ? Isn't it supposed to be an unique key (cherrypy thread names, agent names, request names)?
What's the difference between identifier and doc.thr? This should tell me what is postfix for.

ticoann replied:
This is the special case I had to handle. In wmstats, DataCacheUpdate thread not just run in one threads but all the threads reqmon was deployed. It might be just one reqmon instance now but we can't assume that future either.
In that case I added the postfix with host name to distinguish different thread in different box. (I still need to improve the detection if one thread is not running in one machine though - it won't detect if it never started.)

amaltaro commented on src/python/WMCore/Services/LogDB/LogDB.py:
```diff
@@ -126,5 +127,55 @@ def cleanup(self, thr, backend='local'):
+                    report[thr]['ts'] = ts
+        return report
+    
+    def _append_down_component_detail(self, report, thr, msg, ts=0, state="error"):
```
Instead using ts=0, how about we make this argument mandatory? Otherwise we have no clue when the thread disappeared, do we?

ticoann replied:
If thread crashes we always have the record. if thread never started we won't know. I don't think it makes sense to add time (current time) if the thread never started.

amaltaro reviewed (changes requested):
I'm not sure I followed all these changes (that's why I think we could simplify a few things :)).
Code-wise looks Ok, though I suggest testing it in a clean environment with my suggestions, if you agree.

ticoann commented on src/python/Utils/CherryPyPeriodicTask.py:
```diff
@@ -26,8 +26,19 @@ def __init__(self, config, enableLogDB=False):
         self.setConcurrentTasks(config)
+        self.logDB = None
+        
+        if enableLogDB:
```
I think that is good idea. I will update that.

ticoann commented on src/python/Utils/HeartbeatMonitorBase.py:
```diff
@@ -0,0 +1,36 @@
+    def __init__(self, rest, config):
+        super(HeartbeatMonitorBase, self).__init__(config, enableLogDB=True)
+        
+        from WMCore.Services.WMStats.WMStatsWriter import WMStatsWriter
```
The reason I put this in init method is I am not sure which application has dependence on Utils package. 
Those applications won't be necessary using this class. Or I could move this package some where else (under Services) but I thought this might be the better place.

ticoann commented on src/python/Utils/HeartbeatMonitorBase.py:
```diff
@@ -0,0 +1,36 @@
+        raise NotImplementedError("Need to return list of Class Names for each cherrypy threads to moniotr")
+        
+    def reportToWMStats(self, config):
+        """
```
I will add doctoring. :-)

ticoann commented on src/python/Utils/HeartbeatMonitorBase.py:
```diff
@@ -0,0 +1,36 @@
+        """
+  
+        """
+        from WMCore.Services.WMStats.WMStatsWriter import convertToServiceCouchDoc
```
the same reason above

ticoann commented on src/python/WMCore/GlobalWorkQueue/CherryPyThreads/MonitoringTask.py:
```diff
@@ -4,11 +4,12 @@
 from WMCore.WorkQueue.WorkQueue import globalQueue
 from WMCore.Services.WMStats.WMStatsWriter import WMStatsWriter, convertToServiceCouchDoc
 
+GQ_THREADS = ['CleanUpTask', 'LocationUpdateTask', 'ReqMgrInteractionTask', 'MonitoringTask']
```
I want to eventually have the same as other application HeartbeatMonitor but this one has additional information attached so GQ has a little different structure now.
The reason I explicitly  has the thread list instead of getting from config is that there might be the case we don't want to monitor the thread. (i.e. Heartbeat thread itself or thread running in multiple machines) but I think it wouldn't hurt to monitor everything.  I think it is good idea to get them from config. I will change that.

ticoann commented:
@amaltaro, Hi Alan I updated the patch as you suggested. I tested in vm it seems working fine.
I also moved the package location for CherryPyPeriodicTask.py HeartbeatMonitorBase.py from Utils to WMCore.REST since it is belong to cherrypy server side code. But I am not sure whether that is the right place. If you think there is a better location for that package please let me know.
I still put the LogDB import with in the methods of HeartbeatMonitorBase class, since I am not sure whether some application imports REST packages but not Services packages.

amaltaro commented:
@ticoann I haven't seen any problem with the agent. I think we can merge it before cutting the WMAgent tag.
